wurst/systems/spawns/ResourceSpawns.wurst

```diff
@@ -9,74 +9,122 @@ import ErrorHandling
 
-    construct(int itemsSpawnCount, int animalsSpawnCount, LinkedList<WeightedSpawnRegion> spawnRegionsList)
+    construct(int itemsSpawnCount, int animalsSpawnCount, Pool<rect> spawnRegionsPool)
         this.itemsSpawnCount = itemsSpawnCount
         this.animalsSpawnCount = animalsSpawnCount
-        this.spawnRegionsList = spawnRegionsList
-        for r in this.spawnRegionsList
-            totalRegionsWeight += r.weight
+        this.spawnRegionsPool = spawnRegionsPool
         islandSpawnList.push(this)
 
+    function randomSpawnPoint() returns vec2
+        rect spawnRect = this.spawnRegionsPool.get()
+        while true
+            vec2 spawnPoint = spawnRect.randomPoint()
+            if spawnPoint.isTerrainLand() and spawnPoint.isTerrainWalkable()
+                return spawnPoint
+
+        error("ERROR: randomSpawnPoint() returned null!")
+        return vec2(0,0)
+
     function spawnItemFromPool()
         if this.itemPool == null
             error("ERROR: spawnItemFromPool tried to spawn without having a generated pool!")
             return
 
-        rect spawnRect = pickWeightedSpawnRect()
-        while true
-            vec2 spawnPoint = spawnRect.randomPoint()
-            if spawnPoint.isTerrainLand() and spawnPoint.isTerrainWalkable()
-                let spawnedItem = PlaceRandomItem(itemPool, spawnPoint.x, spawnPoint.y)
+        let spawnPoint = randomSpawnPoint()
+        let spawnedItem = PlaceRandomItem(itemPool, spawnPoint.x, spawnPoint.y)
 
-                //Keep track of spawned total & spawned item type total to balance out spawning
-                this.spawnedTotal += 1
-                let spawnedType = spawnedItem.getTypeId()
-                if not this.spawnsPerType.has(spawnedType)
-                    this.spawnsPerType.put(spawnedType, 1)
-                else
-                    this.spawnsPerType.put(spawnedType, this.spawnsPerType.get(spawnedType) + 1)
+        //Keep track of spawned total & spawned item type total to balance out spawning
+        this.itemsSpawnedTotal += 1
+        let spawnedType = spawnedItem.getTypeId()
+        if not this.spawnsPerType.has(spawnedType)
+            this.spawnsPerType.put(spawnedType, 1)
+        else
+            this.spawnsPerType.put(spawnedType, this.spawnsPerType.get(spawnedType) + 1)
 
-                break
+    function spawnAnimalFromPool()
+        if this.animalPool == null
+            error("ERROR: spawnAnimalFromPool tried to spawn without having a generated pool!")
+            return
+
+        let spawnPoint = randomSpawnPoint()
+
+        let id = this.animalPool.get()
+        switch id
+            case UNIT_ELK
+                new Elk(spawnPoint)
+            case UNIT_SNAKE
+                new Snake(spawnPoint)
+            case UNIT_JUNGLE_WOLF
+                new Wolf(spawnPoint)
+            case UNIT_JUNGLE_BEAR
+                new Bear(spawnPoint)
+            case UNIT_PANTHER
+                new Panther(spawnPoint)
```

Comment:
How does the syntax work with this exactly since those are different classes? Dont you need to cast them when calling anyway?

Reply:
```
let animalPool = new Pool<Function<vec2, UnitEntity>> // or new Pool<VoidFunction<vec2>>
  ..add(pos -> new Elk(pos), 1)
  ..add(pos -> new Snake(pos), 1)
  ..add(pos -> new JungleWolf(pos), 1)
  // ...

let spawner = animalPool.random()
spawner.call(spawnPoint)
```

You shouldn't need to cast them because all they are doing is accepting a `vec2` and performing some action on it

Reply:
Function<T> doesnt take vec2 as an input type

Reply:
Updated